src/hooks.ts

```diff
@@ -0,0 +1,314 @@
+    }
+    return `/* tslint:disable */
+/* eslint-disable */
+// @ts-nocheck
+${hash ? `/* ${hash} */\n` : ""}
+${allImports.join("\n")}
+${typeText || ""}
+
+${docTextComment}
+${nodeStatement}
+(node as any).hash = '${sourceHash}';
+
+export default node;
+
+${allHooks.join("\n")}
+`;
+  };
+
+  return {
+    ...relayCompilerLanguageTypescript(),
+    formatModule,
+  };
+}
+
+function capitalize(str: string) {
+  return `${str[0].toUpperCase()}${str.slice(1)}`;
+}
+
+interface CompilerMeta {
+  derivedFrom?: string;
+  connection?: unknown[];
+  refetch?: {
+    connection: unknown;
+    operation: string;
+    fragmentPathInResult: unknown[];
+    identifierField: unknown;
+  };
+}
+
+function makeImport(idents: string[], from: string) {
+  return `import { ${Array.from(new Set(idents))
+    .sort()
+    .join(", ")} } from "${from}";`;
+}
+
+function makeFragmentBlock(name: string) {
+  const n = capitalize(name);
+
+  // NOTE: These declares ensure that the type of the returned data is:
+  //   - non-nullable if the provided ref type is non-nullable
+  //   - nullable if the provided ref type is nullable
+  //   - array of non-nullable if the provided ref type is an array of
+  //     non-nullable refs
+  //   - array of nullable if the provided ref type is an array of nullable refs
+
+  return `export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: TKey): Required<TKey>[" $data"]
+export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: TKey | null): Required<TKey>[" $data"] | null
+export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: ReadonlyArray<TKey>): ReadonlyArray<Required<TKey>[" $data"]>
+export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: ReadonlyArray<TKey | null>): ReadonlyArray<Required<TKey>[" $data"] | null>
+export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: ReadonlyArray<TKey> | null): ReadonlyArray<Required<TKey>[" $data"]> | null
+export function use${n}Fragment<TKey extends ${n}$key>(fragmentRef: ReadonlyArray<TKey | null> | null): ReadonlyArray<Required<TKey>[" $data"] | null> | null
+export function use${n}Fragment(fragmentRef: any) {
+  return useFragment(node, fragmentRef)
+}`;
+}
+
+function makeRefetchableFragmentBlock(name: string, operation: string) {
+  const n = capitalize(name);
+  return `export function useRefetchable${n}Fragment<TKey extends ${n}$key>(fragmentRef: TKey): [Required<TKey>[" $data"], RefetchFnDynamic<${operation}, ${n}$key>]
+export function useRefetchable${n}Fragment<TKey extends ${n}$key>(fragmentRef: TKey | null): [Required<TKey>[" $data"] | null, RefetchFnDynamic<${operation}, ${n}$key | null>]
+export function useRefetchable${n}Fragment(fragmentRef: any) {
+  return useRefetchableFragment(node, fragmentRef)
+}`;
+}
+
+function makePaginationFragmentBlock(name: string, operation: string) {
+  const n = capitalize(name);
+
+  // Note: It'd be nice if react-relay exported this type for us
```

Comment:
Can it be deeply imported? https://github.com/DefinitelyTyped/DefinitelyTyped/blob/fc5c33122aa1167f3dbbd7f74ef6a6d5a663f9cb/types/react-relay/relay-hooks/usePaginationFragment.d.ts#L6

Reply:
Those DefinitelyTyped files are lying a bit 😆 - I suppose we could do `import type { usePaginationFragmentHookType } from 'react-relay/relay-hooks/usePaginationFragment` but that's technically not a valid import path from the js.

Reply:
Ah, yep. Makes sense.